### ISSUE TITLE
Strange behavior of similar for sparse arrays

### ISSUE THREAD

**Issue by bkamins** [JuliaLang/julia#42000]
We have the following:

```
julia> x = spzeros(1)
1-element SparseVector{Float64, Int64} with 0 stored entries

julia> similar(x)
1-element SparseVector{Float64, Int64} with 0 stored entries

julia> x[1] = 1
1

julia> x
1-element SparseVector{Float64, Int64} with 1 stored entry:
  [1]  =  1.0

julia> similar(x)
1-element SparseVector{Float64, Int64} with 1 stored entry:
  [1]  =  6.36599e-314
```

Which seems strange to me as `similar` should produce an uninitialized array, while in the second example it is initialized as I would consider sparse array uninitialized if it had 0 stored entries.

CC @nalimilan

**Comment by martinholters** [comment]
I don't think we make any guarantees to what an (in the sense of `similar`) uninitialized (sparse or otherwise) array gets initialized. For sparse arrays it currently preserves the sparsity pattern with the expectation that it will often be used to allocate the output of an operation which preserves the sparsity pattern of its input. (We might even rely on that internally, but I also would not expect this behavior to be guaranteed.)

**Comment by bkamins** [comment]
Yes - I can see that this is the current pattern.

If this is intended then maybe it should be documented as in the cases when one wants truly uninitialized sparse array one should not use `similar`? (the point is that what `similar` currently does potentially allocates a lot)

**Comment by nalimilan** [comment]
I guess it would make sense to document this as it affects performance and memory use. But note that sparse arrays never have uninitialized entries AFAIK: all non-stored entries are treated as zeros.
```
julia> x = sparse(Int[], Int[], BigFloat[], 2, 3)
2×3 SparseMatrixCSC{BigFloat, Int64} with 0 stored entries:
  ⋅    ⋅    ⋅ 
  ⋅    ⋅    ⋅ 

julia> isassigned(x, 1, 1)
true

julia> x[1, 1]
0.0
```
